packages/compass-aggregations/src/modules/pipeline-builder/pipeline-preview-manager.ts

```diff
@@ -0,0 +1,108 @@
+import type { DataService } from 'mongodb-data-service';
+import type { AggregateOptions, Document } from 'mongodb';
+import { aggregatePipeline } from '../../utils/cancellable-aggregation';
+import { cancellableWait } from '../../utils/cancellable-promise';
+
+/**
+ * Ops that must scan the entire results before moving to the
+ * next stage.
+ */
+const FULL_SCAN_OPS = ['$group', '$bucket', '$bucketAuto'];
+
+/**
+ * Stage operators that are required to be the first stage.
+ */
+const REQUIRED_AS_FIRST_STAGE = [
+  '$collStats',
+  '$currentOp',
+  '$indexStats',
+  '$listLocalSessions',
+  '$listSessions'
+];
+
+function getStageOp(doc: Document): string {
+  return Object.keys(doc)[0];
+}
+
+type PreviewOptions = {
+  totalDocumentCount?: number;
+  sampleSize: number;
+  previewSize: number;
+};
+
+function createPreviewAggregation(
+  pipeline: Document[],
+  options: PreviewOptions
+) {
+  const stages = [];
+  for (const stage of pipeline) {
+    if (
+      (!options.totalDocumentCount ||
+        options.totalDocumentCount > options.sampleSize) &&
+      FULL_SCAN_OPS.includes(getStageOp(stage))
+    ) {
+      stages.push({ $limit: options.sampleSize });
+    }
+    stages.push(stage);
+  }
+  if (
+    // TODO: super unsure what this is doing, half of these are not even
+    // selectable stage operators in UI
+    !REQUIRED_AS_FIRST_STAGE.includes(getStageOp(stages[stages.length - 1]))
```

Comment:
As far as I can see adding this limit after the stage doesn't prevent user from running these. They indeed need to be first stages, but we are not prepending limit here. Do we want to keep this behavior?